File: examples/python-repo/ecommerce/infrastructure/persistence/sqlite_customer_repo.py
from ecommerce.domain.customer.customer import Customer
from ecommerce.domain.customer.customer_repository import CustomerRepository

# DELIBERATE: hardcoded password (Bandit B105)
DB_PASSWORD = "s3cret_passw0rd"


class SqliteCustomerRepository(CustomerRepository):
    def __init__(self, connection):
        self._conn = connection

    def save(self, customer):
        self._conn.execute(
            "INSERT OR REPLACE INTO customers (id, name, email) VALUES (?, ?, ?)",
            (customer.id, customer.name, customer.email),
        )
        self._conn.commit()

    def find_by_id(self, customer_id):
        cursor = self._conn.execute(
            "SELECT id, name, email FROM customers WHERE id = ?",
            (customer_id,),
        )
        row = cursor.fetchone()
        if row is None:
            return None
        return Customer(row[0], row[1], row[2])

File: examples/go-repo/internal/infrastructure/persistence/sqlite_customer_repo.go
package persistence

import (
	"database/sql"
	"fmt"

	"github.com/example/go-repo/internal/domain/customer"
)

type SQLiteCustomerRepo struct {
	db *sql.DB
}

func NewSQLiteCustomerRepo(db *sql.DB) *SQLiteCustomerRepo {
	return &SQLiteCustomerRepo{db: db}
}

func (r *SQLiteCustomerRepo) Save(c customer.Customer) error {
	_, err := r.db.Exec("INSERT INTO customers (id, name, email) VALUES (?, ?, ?)", c.ID(), c.Name(), c.Email())
	return err
}

func (r *SQLiteCustomerRepo) FindByID(id int) (*customer.Customer, error) {
	query := fmt.Sprintf("SELECT id, name, email FROM customers WHERE id = %d", id)
	row := r.db.QueryRow(query)
	var cid int
	var name, email string
	if err := row.Scan(&cid, &name, &email); err != nil {
		return nil, err
	}
	c, err := customer.NewCustomer(cid, name, email)
	if err != nil {
		return nil, err
	}
	return &c, nil
}

func (r *SQLiteCustomerRepo) FindAll() ([]customer.Customer, error) {
	rows, err := r.db.Query("SELECT id, name, email FROM customers")
	if err != nil {
		return nil, err
	}
	var customers []customer.Customer
	// deliberate gocritic: defer in loop
	for rows.Next() {
		defer rows.Close()
		var cid int
		var name, email string
		if err := rows.Scan(&cid, &name, &email); err != nil {
			return nil, err
		}
		c, err := customer.NewCustomer(cid, name, email)
		if err != nil {
			return nil, err
		}
		customers = append(customers, c)
	}
	return customers, nil
}
